src/client/automation/visible-element-automation.ts

```diff
@@ -153,28 +158,89 @@ export default class VisibleElementAutomation extends SharedEventEmitter {
             });
     }
 
-    private _getElementOffset (): { offsetX: number; offsetY: number } {
+    private _getElementOffset (): PointOptions {
         const defaultOffsets = getOffsetOptions(this.element);
 
-        let { offsetX, offsetY } = this.options;
+        const { offsetX, offsetY } = this.options;
 
-        offsetX = offsetX || offsetX === 0 ? offsetX : defaultOffsets.offsetX;
-        offsetY = offsetY || offsetY === 0 ? offsetY : defaultOffsets.offsetY;
+        const y = offsetY || offsetY === 0 ? offsetY : defaultOffsets.offsetY;
+        const x = offsetX || offsetX === 0 ? offsetX : defaultOffsets.offsetX;
 
-        return { offsetX, offsetY };
+        return { x, y };
+    }
+
+    private async _isTargetElement ( element: HTMLElement, expectedElement: HTMLElement | null): Promise<boolean> {
+        let isTarget = !expectedElement || element === expectedElement || element === this.element;
+
+        if (!isTarget && element) {
+            // NOTE: perform an operation with searching in dom only if necessary
+            isTarget = await this._contains(this.element, element);
+        }
+
+        return isTarget;
+    }
+
+    private async _getAvailableOffset (expectedElement: HTMLElement | null, pointOptions: PointOptions, deep = 2): Promise<PointOptions | null> {
+        let screenPoint = await getAutomationPoint(this.element, { ...pointOptions });
+        let clientPoint = await screenPointToClient(this.element, screenPoint);
+        let element     = await getElementFromPoint(clientPoint, this.window, expectedElement as HTMLElement);
+        let isTarget    = element && await this._isTargetElement(element, expectedElement);
+
+        if (element && isTarget)
+            return pointOptions;
+
+        const stepX = pointOptions.x / deep;
```

Comment:
It's difficult to read and understand.
I propose to rewrite the code in the following way:
```js
// calculate all checked points before
const checkedPoints = this._getCheckedPoints(...);

// perform checking the target element
do {
      // These function can be moved to the `findTargetElement()` function.
      screenPoint = await getAutomationPoint(this.element, { x, y });
       clientPoint = await screenPointToClient(this.element, screenPoint);
       element     = await getElementFromPoint(clientPoint, this.window, expectedElement as HTMLElement);
}

while(await this._isTargetElement(element, expectedElement));
```

Also, the point calculation logic can be easily covered by client tests.